shared/shared.go

```diff
@@ -176,6 +187,35 @@ func GetTmpDir() string {
 	return filepath.Join(getDataDir(), "tmp")
 }
 
+func WriteIPSalt(w io.Writer) (int, error) {
+	return w.Write(ipSaltValue)
+}
+
+func loadIPSalt(config *toml.Tree, dataDir string) ([]byte, error) {
+	ipSaltMethod := "auto"
+	if key := config.Get("ip_salt"); key != nil {
+		ipSaltMethod = key.(string)
+	}
+
+	ipSaltPath := ipSaltMethod
+	switch ipSaltMethod {
+	case "disabled":
+		return nil, nil
+	case "auto":
+		ipSaltPath = filepath.Join(dataDir, "ip_salt")
+		f, err := os.OpenFile(ipSaltPath, os.O_RDWR|os.O_CREATE|os.O_EXCL, 0o644)
+		if err == nil {
+			// First time, create random data
+			defer f.Close()
+			if _, err := io.CopyN(f, rand.Reader, 16); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return ioutil.ReadFile(ipSaltPath)
```

Comment:
If we just created the salt, then this line will try to read from that file before it's even closed. It would be better to just return the salt after creating it on line 210, if there's no error.